File: IAIso-v5.0/core/iaiso-csharp/src/Iaiso.Core/BoundedExecution.cs
using System;
using System.Collections.Generic;
using System.Threading;
using Iaiso.Audit;

namespace Iaiso.Core;

/// <summary>Options for <see cref="BoundedExecution.Start"/>.</summary>
public sealed class BoundedExecutionOptions
{
    public string? ExecutionId { get; set; }
    public PressureConfig Config { get; set; } = PressureConfig.Defaults();
    public ISink? AuditSink { get; set; }
    public IClock? Clock { get; set; }
    public IClock? TimestampClock { get; set; }
}

/// <summary>
/// High-level execution facade. Composes a <see cref="PressureEngine"/>
/// with an audit sink and lifecycle management.
/// </summary>
/// <remarks>
/// Use <see cref="Run(BoundedExecutionOptions, Action{BoundedExecution})"/>
/// for the closure style with automatic cleanup, or <see cref="Start"/>
/// + <c>using</c> for explicit lifecycle control:
/// <code>
/// using var exec = BoundedExecution.Start(opts);
/// exec.RecordToolCall("search", 500);
/// </code>
/// </remarks>
public sealed class BoundedExecution : IDisposable
{
    private readonly PressureEngine _engine;
    private readonly ISink _auditSink;
    private readonly IClock _timestampClock;
    private int _closed; // 0 == open, 1 == closed (atomic transitions)

    private BoundedExecution(BoundedExecutionOptions opts)
    {
        var execId = opts.ExecutionId;
        if (string.IsNullOrEmpty(execId))
        {
            execId = "exec-" + System.Diagnostics.Stopwatch.GetTimestamp().ToString("x");
        }
        var clk = opts.Clock ?? WallClock.Instance;
        var tsClk = opts.TimestampClock ?? clk;
        _auditSink = opts.AuditSink ?? NullSink.Instance;
        _timestampClock = tsClk;
        _engine = new PressureEngine(opts.Config, new EngineOptions
        {
            ExecutionId = execId,
            AuditSink = _auditSink,
            Clock = clk,
            TimestampClock = tsClk,
        });
    }

    /// <summary>Construct a <see cref="BoundedExecution"/>. Caller MUST <see cref="Dispose"/>.</summary>
    public static BoundedExecution Start(BoundedExecutionOptions opts) => new(opts);

    /// <summary>Run a closure inside a bounded execution; closes on exit.</summary>
    public static void Run(BoundedExecutionOptions opts, Action<BoundedExecution> body)
    {
        bool errored = false;
        BoundedExecution? exec = null;
        try
        {
            exec = Start(opts);
            body(exec);
        }
        catch
        {
            errored = true;
            throw;
        }
        finally
        {
            exec?.CloseWith(errored);
        }
    }

    public PressureEngine Engine => _engine;
    public PressureSnapshot Snapshot() => _engine.Snapshot();

    /// <summary>Account for tokens with an optional tag.</summary>
    public StepOutcome RecordTokens(long tokens, string? tag = null) =>
        _engine.Step(new StepInput { Tokens = tokens, Tag = tag });

    /// <summary>Account for a single tool invocation.</summary>
    public StepOutcome RecordToolCall(string name, long tokens) =>
        _engine.Step(new StepInput { Tokens = tokens, ToolCalls = 1, Tag = name });

    /// <summary>General step accounting.</summary>
    public StepOutcome RecordStep(StepInput work) => _engine.Step(work);

    /// <summary>Pre-check the engine state without advancing it.</summary>
    public StepOutcome Check() => _engine.Lifecycle switch
    {
        Lifecycle.Locked => StepOutcome.Locked,
        Lifecycle.Escalated => StepOutcome.Escalated,
        _ => StepOutcome.Ok,
    };

    public PressureSnapshot Reset() => _engine.Reset();

    public void Dispose() => CloseWith(false);

    internal void CloseWith(bool errored)
    {
        if (Interlocked.CompareExchange(ref _closed, 1, 0) != 0) return;
        var snap = _engine.Snapshot();
        var data = new SortedDictionary<string, object?>
        {
            ["final_pressure"] = snap.Pressure,
            ["final_lifecycle"] = snap.Lifecycle.Wire(),
            ["exception"] = errored ? "error" : null,
        };
        _auditSink.Emit(new Event(_engine.ExecutionId, "execution.closed",
            _timestampClock.Now(), data));
    }
}

File: IAIso-v5.0/core/iaiso-php/src/Core/BoundedExecution.php
<?php

declare(strict_types=1);

namespace IAIso\Core;

use IAIso\Audit\Event;
use IAIso\Audit\NullSink;
use IAIso\Audit\Sink;

/**
 * High-level execution facade. Composes a {@see PressureEngine} with
 * an audit sink and lifecycle management.
 *
 * <p>Use {@see BoundedExecution::run()} for the closure style with
 * automatic cleanup, or {@see BoundedExecution::start()} + manual
 * {@see BoundedExecution::close()} for explicit lifecycle control.
 */
final class BoundedExecution
{
    private readonly PressureEngine $engine;
    private readonly Sink $auditSink;
    private readonly Clock $timestampClock;
    private bool $closed = false;

    private function __construct(BoundedExecutionOptions $opts)
    {
        $execId = $opts->executionId !== null && $opts->executionId !== ''
            ? $opts->executionId
            : 'exec-' . bin2hex(random_bytes(6));
        $cfg = $opts->config ?? PressureConfig::defaults();
        $clock = $opts->clock ?? WallClock::instance();
        $tsClock = $opts->timestampClock ?? $clock;
        $this->auditSink = $opts->auditSink ?? NullSink::instance();
        $this->timestampClock = $tsClock;
        $this->engine = new PressureEngine($cfg,
            new EngineOptions($execId, $this->auditSink, $clock, $tsClock));
    }

    /** Construct a {@see BoundedExecution}. The caller MUST {@see close()} it. */
    public static function start(BoundedExecutionOptions $opts): self
    {
        return new self($opts);
    }

    /**
     * Run a closure inside a bounded execution; closes on exit.
     *
     * @param callable(self):void $body
     */
    public static function run(BoundedExecutionOptions $opts, callable $body): void
    {
        $exec = self::start($opts);
        $errored = false;
        try {
            $body($exec);
        } catch (\Throwable $e) {
            $errored = true;
            throw $e;
        } finally {
            $exec->closeWith($errored);
        }
    }

    public function getEngine(): PressureEngine     { return $this->engine; }
    public function snapshot(): PressureSnapshot    { return $this->engine->snapshot(); }

    /** Account for tokens with an optional tag. */
    public function recordTokens(int $tokens, ?string $tag = null): StepOutcome
    {
        return $this->engine->step(new StepInput(tokens: $tokens, tag: $tag));
    }

    /** Account for a single tool invocation. */
    public function recordToolCall(string $name, int $tokens = 0): StepOutcome
    {
        return $this->engine->step(new StepInput(tokens: $tokens, toolCalls: 1, tag: $name));
    }

    /** General step accounting. */
    public function recordStep(StepInput $work): StepOutcome
    {
        return $this->engine->step($work);
    }

    /** Pre-check the engine state without advancing it. */
    public function check(): StepOutcome
    {
        return match ($this->engine->getLifecycle()) {
            Lifecycle::Locked    => StepOutcome::Locked,
            Lifecycle::Escalated => StepOutcome::Escalated,
            default              => StepOutcome::Ok,
        };
    }

    public function reset(): PressureSnapshot
    {
        return $this->engine->reset();
    }

    /** Close the execution, emitting {@code execution.closed}. Idempotent. */
    public function close(): void
    {
        $this->closeWith(false);
    }

    private function closeWith(bool $errored): void
    {
        if ($this->closed) return;
        $this->closed = true;
        $snap = $this->engine->snapshot();
        $this->auditSink->emit(new Event(
            $this->engine->getExecutionId(),
            'execution.closed',
            $this->timestampClock->now(),
            [
                'final_pressure'  => $snap->pressure,
                'final_lifecycle' => $snap->lifecycle->value,
                'exception'       => $errored ? 'error' : null,
            ],
        ));
    }

    public function __destruct()
    {
        // Defensive: PHP's GC will fire this if the user forgets to close().
        $this->closeWith(false);
    }
}
